internal/clients/azuread.go

```diff
@@ -93,3 +115,84 @@ func configureTFProviderMeta(ctx context.Context, ps *terraform.Setup, p schema.
 	ps.Meta = p.Meta()
 	return nil
 }
+
+func spAuth(ctx context.Context, pc *v1beta1.ProviderConfig, ps *terraform.Setup, client client.Client) error {
+	data, err := xpresource.CommonCredentialExtractor(ctx, pc.Spec.Credentials.Source, client, pc.Spec.Credentials.CommonCredentialSelectors)
+	if err != nil {
+		return errors.Wrap(err, errExtractCredentials)
+	}
+	data = []byte(strings.TrimSpace(string(data)))
+	azureCreds := map[string]string{}
+	if err := json.Unmarshal(data, &azureCreds); err != nil {
+		return errors.Wrap(err, errUnmarshalCredentials)
+	}
+	// set credentials configuration
+	ps.Configuration[keySubscriptionID] = azureCreds[keyAzureSubscriptionID]
+	ps.Configuration[keyTenantID] = azureCreds[keyAzureTenantID]
+	ps.Configuration[keyClientID] = azureCreds[keyAzureClientID]
+	ps.Configuration[keyClientSecret] = azureCreds[keyAzureClientSecret]
+	if clientCert, ok := azureCreds[keyAzureClientCert]; ok {
+		ps.Configuration[keyClientCert] = clientCert
+		if clientCertPass, passwordOk := azureCreds[keyAzureClientCertPass]; passwordOk {
+			ps.Configuration[keyClientCertPassword] = clientCertPass
+		}
+	}
+	if pc.Spec.SubscriptionID != nil {
+		ps.Configuration[keySubscriptionID] = *pc.Spec.SubscriptionID
+	}
+	if pc.Spec.TenantID != nil {
+		ps.Configuration[keyTenantID] = *pc.Spec.TenantID
+	}
+	if pc.Spec.ClientID != nil {
+		ps.Configuration[keyClientID] = *pc.Spec.ClientID
+	}
+	if pc.Spec.Environment != nil {
+		ps.Configuration[keyEnvironment] = *pc.Spec.Environment
+	}
+	return nil
+}
+
+func msiAuth(pc *v1beta1.ProviderConfig, ps *terraform.Setup) error {
+	if pc.Spec.SubscriptionID == nil || len(*pc.Spec.SubscriptionID) == 0 {
+		return errors.New(errSubscriptionIDNotSet)
+	}
+	if pc.Spec.TenantID == nil || len(*pc.Spec.TenantID) == 0 {
+		return errors.New(errTenantIDNotSet)
+	}
+	ps.Configuration[keySubscriptionID] = *pc.Spec.SubscriptionID
+	ps.Configuration[keyTenantID] = *pc.Spec.TenantID
+	ps.Configuration[keyUseMSI] = "true"
+	if pc.Spec.MSIEndpoint != nil {
+		ps.Configuration[keyMSIEndpoint] = *pc.Spec.MSIEndpoint
+	}
+	if pc.Spec.ClientID != nil {
+		ps.Configuration[keyClientID] = *pc.Spec.ClientID
+	}
+	if pc.Spec.Environment != nil {
+		ps.Configuration[keyEnvironment] = *pc.Spec.Environment
+	}
+	return nil
+}
+
+func oidcAuth(pc *v1beta1.ProviderConfig, ps *terraform.Setup) error {
+	if pc.Spec.SubscriptionID == nil || len(*pc.Spec.SubscriptionID) == 0 {
+		return errors.New(errSubscriptionIDNotSet)
+	}
+	if pc.Spec.TenantID == nil || len(*pc.Spec.TenantID) == 0 {
+		return errors.New(errTenantIDNotSet)
+	}
+	if pc.Spec.ClientID == nil || len(*pc.Spec.ClientID) == 0 {
+		return errors.New(errClientIDNotSet)
```

Comment:
The oidcAuth method should support set  `Environment` variable

Reply:
My goal here was to bring it up to par with `provider-upjet-azure`. This would be an extra feature which wouldn't be supported by the other provider unless implemented too. I prefer having both providers behave the same.